Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+    request: ClientRequest.Stream<Request>,
+    descriptor: MethodDescriptor,
+    serializer: some MessageSerializer<Request>,
+    deserializer: some MessageDeserializer<Response>,
+    handler: @Sendable @escaping (ClientResponse.Stream<Response>) async throws -> ReturnValue
+  ) async throws -> ReturnValue {
+    let (configurationOverrides, interceptors) = try self.storage.withLockedValue { storage in
+      switch storage.state {
+      case .running:
+        return (storage.methodConfigurationOverrides, storage.interceptors.values)
+      case .notStarted:
+        throw ClientError(
+          code: .clientIsNotRunning,
+          message: "Client must be running to make an RPC: call run() first."
+        )
+      case .stopping, .stopped:
+        throw ClientError(
+          code: .clientIsStopped,
+          message: "Client has been stopped. Can't make any more RPCs."
+        )
+      }
+    }
+
+    let applicableConfiguration = self.resolveMethodConfiguration(
+      descriptor: descriptor,
+      clientConfigurations: configurationOverrides
+    )
+
+    return try await ClientRPCExecutor.execute(
+      request: request,
+      method: descriptor,
+      configuration: applicableConfiguration,
+      serializer: serializer,
+      deserializer: deserializer,
+      transport: transport,
+      interceptors: interceptors,
+      handler: handler
+    )
+  }
+
+  private func resolveMethodConfiguration(
+    descriptor: MethodDescriptor,
+    clientConfigurations configurationOverrides: MethodConfigurationRegistry
+  ) -> MethodConfiguration {
+    if let clientOverride = configurationOverrides[descriptor, useDefault: false] {
+      return clientOverride
+    }
+
+    if let transportConfiguration = self.transport.executionConfiguration(forMethod: descriptor) {
+      return transportConfiguration
+    }
+
+    return configurationOverrides[descriptor]
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension GRPCClient {
+  /// A collection of interceptors providing cross-cutting functionality to each accepted RPC.
+  public struct Interceptors: Sendable {
+    private(set) var values: [any ClientInterceptor] = []
+
+    /// Add an interceptor to the server.
+    ///
+    /// The order in which interceptors are added reflects the order in which they are called. The
+    /// first interceptor added will be the first interceptor to intercept each request. The last
+    /// interceptor added will be the final interceptor to intercept each request before calling
+    /// the appropriate handler.
+    ///
+    /// - Parameter interceptor: The interceptor to add.
+    public mutating func add(_ interceptor: some ClientInterceptor) {
+      self.values.append(interceptor)
+    }
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension GRPCClient {
+  /// The execution policy for an RPC.
+  public enum ExecutionPolicy: Hashable, Sendable {
```

Comment:
Can we use a struct for this instead?

Reply:
We discussed this one previously (it's just been renamed from elsewhere)